Sources/TestingInternals/Discovery.cpp

```diff
@@ -248,105 +322,108 @@ static SWTMachHeaderList getMachHeaders(void) {
-  MetadataSectionRange swift5_type_metadata;
-  MetadataSectionRange swift5_typeref;
-  MetadataSectionRange swift5_reflstr;
-  MetadataSectionRange swift5_fieldmd;
-  MetadataSectionRange swift5_assocty;
-  MetadataSectionRange swift5_replace;
-  MetadataSectionRange swift5_replac2;
-  MetadataSectionRange swift5_builtin;
-  MetadataSectionRange swift5_capture;
-  MetadataSectionRange swift5_mpenum;
-  MetadataSectionRange swift5_accessible_functions;
-};
-
 /// A function exported by the Swift runtime that enumerates all metadata
 /// sections loaded into the current process.
 SWT_IMPORT_FROM_STDLIB void swift_enumerateAllMetadataSections(
-  bool (* body)(const MetadataSections *sections, void *context),
+  bool (* body)(const SWTMetadataSections *sections, void *context),
   void *context
 );
 
-template <typename SectionEnumerator>
-static void enumerateTypeMetadataSections(const SectionEnumerator& body) {
-  swift_enumerateAllMetadataSections([] (const MetadataSections *sections, void *context) {
-    const auto& body = *reinterpret_cast<const SectionEnumerator *>(context);
-    MetadataSectionRange section = sections->swift5_type_metadata;
-    if (section.start && section.length > 0) {
-      body(reinterpret_cast<const void *>(section.start), section.length);
-    }
+template <typename ImageEnumerator>
+static void enumerateImages(const ImageEnumerator& body) {
+  swift_enumerateAllMetadataSections([] (const SWTMetadataSections *sections, void *context) {
+    const auto& body = *reinterpret_cast<const ImageEnumerator *>(context);
+    body(*sections);
     return true;
-  }, const_cast<SectionEnumerator *>(&body));
+  }, const_cast<ImageEnumerator *>(&body));
 }
 #endif
 
 #pragma mark -
 
 void swt_enumerateTypes(void *context, SWTTypeEnumerator body, SWTTypeNameFilter nameFilter) {
-  enumerateTypeMetadataSections([=] (const void *section, size_t size) {
-    auto records = reinterpret_cast<const SWTTypeMetadataRecord *>(section);
-    size_t recordCount = size / sizeof(SWTTypeMetadataRecord);
-
-    for (size_t i = 0; i < recordCount; i++) {
-      const auto& record = records[i];
-
-      auto contextDescriptor = record.getContextDescriptor();
-      if (!contextDescriptor) {
-        // This type metadata record is invalid (or we don't understand how to
-        // get its context descriptor), so skip it.
-        continue;
-      } else if (contextDescriptor->isGeneric()) {
-        // Generic types cannot be fully instantiated without generic
-        // parameters, which is not something we can know abstractly.
-        continue;
-      }
+  enumerateImages([=] (const SWTMetadataSections& sections) {
+    // If this image has a tests section, then it should be used instead of the
+    // image's type metadata section.
+    if (sections.version >= SWT_METADATA_SECTION_MINIMUM_VERSION_WITH_TESTS
+        && sections.swift5_tests.start != 0
+        && sections.swift5_tests.length > 0) {
+      return; // continue
+    }
```

Comment:
Personally, I think that splitting this up might be a bit nicer:

```c++
if (sections.version < SWT_METADATA_SECTION_MINIMUM_VERSION_WITH_TESTS)
  return;

const auto &swift5_tests = sections.swift5_tests;
if (swift5_tests.start == 0 || swift5_tests.length <= 0)
  return;
```

Reply:
I can do that, sure.

Reply:
Well, not exactly that, as we want to fall back to the path that looks at type metadata. But I'll split it up regardless.